### PR TITLE
anuko-time-tracker: init at 1.22.19.5806

### PR DESCRIPTION
**Pull request by michaelshmitty**
###### Description of changes

[Anuko Time Tracker](https://github.com/anuko/timetracker) is a simple, easy to use, open source time tracking system written in PHP.

###### Things done
- Packaging steps
  - [x] Package Anuko Time Tracker
  - [x] Create NixOS module for a service
  - [x] Add release note changes
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes **(no longer required)**
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by michaelshmitty** [comment]
This is my first contribution to nixpkgs. Let me know if something doesn't check out and what I can do to fix it.

**Comment by michaelshmitty** [comment]
I intend to create a NixOS module to run this webapp as a service. That seems a bit more daunting and will take some more time to learn.

File: pkgs/servers/web-apps/anuko/default.nix
Comment (Janik-Haag) [inline pkgs/servers/web-apps/anuko/default.nix:33]: The with pattern is discouraged, you should try to avoid it and do something like the suggestion bellow, you can read more about it here https://github.com/NixOS/nixpkgs/pull/209325
``` suggestion
meta = {
    description = "Simple, easy to use, open source time tracking system";
    homepage = "https://github.com/anuko/timetracker/";
    license = lib.licenses.sspl;
    maintainers = with lib.maintainers; [ michaelshmitty ];
```

**Review by Janik-Haag — CHANGES_REQUESTED** [review]

File: pkgs/servers/web-apps/anuko/default.nix
Comment (michaelshmitty) [reply pkgs/servers/web-apps/anuko/default.nix:33]: @Janik-Haag I'll fix it, thanks for the feedback! I'm using existing packages in nixpkgs as a base, but it's hard to know what I'm copying over is good to begin with :).
If there are more places where these kind of discouragements are documented, please let me know!

Comment (michaelshmitty) [reply pkgs/servers/web-apps/anuko/default.nix:33]: > If there are more places where these kind of discouragements are documented, please let me know!

@Janik-Haag nevermind, I just found this https://nix.dev/anti-patterns/language

**Review by Janik-Haag — COMMENTED** [review]
Looks fine so far, but your missing the release note changes, you can look at the release file here nixos/doc/manual/release-notes/[rl-2305.section.md](http://rl-2305.section.md/) and here is the manual describing the contribution steps. https://nixos.org/manual/nixos/unstable/#chap-contributing

It would also be nice if you could add a Service, you can find examples for that in nixos/modules/services/
I would recommand you look at some other php service

**Comment by michaelshmitty** [comment]
Hi @Janik-Haag , I've added a NixOS module for the Anuko Time Tracker service, a test for the module and release notes. Would you mind having another look at this PR? Many thanks!

As I expected the service took way longer to implement than the package, but I learned a lot from what was already there in `nixos/modules/services/web-apps`. If you can, please check if I haven't followed any known anti-patterns again.

Now I am curious on how I can use this service in my NixOS server's system configuration before this PR is merged. And even so, my servers system configuration use the latest stable release (22.11), but this PR is based on master.
I use flakes by the way.

So I figure I could add another input to my flake, pointing to my fork of nixpkgs, something like:
```
inputs.my-nixpkgs.url = "github:michaelshmitty/nixpkgs/anuko-timetracker";
```
But then I am not sure how I can use the service in my nixos configuration. Would I just do:
```
my-nixpkgs.services.anuko-time-tracker.enable = true;
```
or something like that? If that even works, how will it handle the fact that my entire server setup uses the nixpkgs input from the 22.11 branch, but this would then reference the master branch? I will experiment a bit by myself, but I'm a bit confused about how that is resolved, if it is even possible. If you or anyone else has any experience with that, thanks!

**Comment by michaelshmitty** [comment]
Thanks to suggestions from @robbevp I managed to create an overlay that brings the anuko-time-tracker from my nixpkgs fork into my flake, as well as import the anuko-time-tracker NixOS module for the service from my nixpkgs fork into the configuration of one of my hosts. At the end I struggled quite a bit with the fact that it would not honor the allowUnfree config setting, but I managed to get that working too:

First I added my nixpkgs fork on the anuko-time-tracker branch as an input to my servers configuration `flake.nix`:
```
{
  ...
  inputs.my-nixpkgs.url = "github:michaelshmitty/nixpkgs/anuko-timetracker";
  ...
}
```
Then I had to re-import my-nixpkgs to be able to set the config to allowUnfree = true; Otherwise it would not evaluate the anuko-time-tracker package even though my host configuration has nixpkgs.config.allowUnfree = true; set. This tripped me up for quite a while. I managed to fix it with some help from SO like this:
```
outputs = { self, nixpkgs, my-nixpkgs, ... }:
  let
    ...
    my-nixpgs-with-anuko = import my-nixpkgs {
        system = "x86_64-linux";
        config.allowUnfree = true;
    };
  in
  {
    nixosConfigurations = {
      some-hostname = nixpkgs.lib.nixosSystem {
          system = "x86_64-linux";
          modules = [
            # Bring in the package:
            ({ pkgs, ... }: {
              nixpkgs.overlays = [
                (self: super: {
                  anuko-time-tracker = my-nixpgs-with-anuko.anuko-time-tracker;
                })
              ];
            })
            # Bring in the Nixos module for the service:
            "${my-nixpkgs}/nixos/modules/services/web-apps/anuko-time-tracker.nix"
            ./machines/some-hostname
          ];
        };
    };
    ... 
  }
```

Then in `./machines/some-hostname/default.nix` I was finally able to do:
```
services.anuko-time-tracker.enable = true;
```

If anyone has suggestions on how this can be improved, I'm all ears. If not, I hope it can help someone else struggling with this.

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (AndersonTorres) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:32]: Why are you repackaging?

Comment (michaelshmitty) [reply nixos/modules/services/web-apps/anuko-time-tracker.nix:32]: @AndersonTorres 
Hey sorry I didn't reply earlier. Even after reviewing github notification settings multiple times, I still seem to miss important ones.

I'm repackaging because the output that gets hosted needs the config file linked into $out/WEB_INF/config.php and it needs a symlink writable location linked into $out/WEB-INF/templates_c
I copied this behavior / workaround from some other PHP webapp NixOS module. Please tell me if this can be improved, happy to learn, thanks!

Comment (michaelshmitty) [reply nixos/modules/services/web-apps/anuko-time-tracker.nix:32]: @AndersonTorres I had to refresh my memory but this is where I got the inspiration:
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/cloudlog.nix#L44

File: maintainers/maintainer-list.nix
Comment (AndersonTorres) [inline maintainers/maintainer-list.nix:9863]: reword this commit:

`maintainers: add michaelshmitty`

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:346]:
```suggestion
      tmpfiles.rules = [
        "d ${cfg.dataDir} 0750 ${cfg.user} ${config.services.nginx.group} - -"
      ];
```

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:337]: We are using tmpfiles.d to do this

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:290]: We can't use postgres?

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:259]: Whats the reason behind this?

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:197]: We should use a password file if possible

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:188]: I don't think we can set a good default here. It could also be 587

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:170]: We should not send mails from an address that is probably not belonging to that instance by default.

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:124]:
```suggestion
        any virtualhost.
```
this should be named domain

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:57]: There is mkPackageOption

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:34]:
```suggestion
    inherit (src) version;
```

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:44]:
```suggestion
      ln -fs ${cfg.dataDir}/templates_c $out/WEB-INF/templates_c
```

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:38]:
```suggestion
      mkdir -p $out
      lndir ./ $out/
```

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:35]:
```suggestion
    src = pkgs.anuko-time-tracker;
    nativeBuildInputs = [ xorg.lndir ];
```

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/anuko-time-tracker.nix:9]: This should use the package option

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (michaelshmitty) [reply nixos/modules/services/web-apps/anuko-time-tracker.nix:259]: This is taken from https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/cloudlog.nix#L300
I think the reasoning is that if you let the nixos module create the database locally, you don't need to supply a password. Not sure what the reasoning behind that is either.
If you know a better (PHP based) web app example in nixpkgs that I should take inspiration from let me know!

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (michaelshmitty) [reply nixos/modules/services/web-apps/anuko-time-tracker.nix:197]: I agree, I'll fix this.

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (michaelshmitty) [reply nixos/modules/services/web-apps/anuko-time-tracker.nix:188]: Although I agree with your comment, for the defaults I have been using [the upstream project's example documentation](https://www.anuko.com/time-tracker/install-guide/mail.htm). So what's better here? Leave their instructions or no default?

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (michaelshmitty) [reply nixos/modules/services/web-apps/anuko-time-tracker.nix:170]: Yeah, you're right, but same as above, these are [the upstream project's defaults](https://www.anuko.com/time-tracker/install-guide/mail.htm). Maybe I should revise this.

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (SuperSandro2000) [reply nixos/modules/services/web-apps/anuko-time-tracker.nix:292]: If upstream supports postgres, we want to default to that but if it doesn't than thats unfortunate.

**Review by Janik-Haag — COMMENTED** [review]
@michaelshmitty can you do a rebase and address the open suggestions so we can get this merged?

**Comment by michaelshmitty** [comment]
@Janik-Haag sure, I'll revisit this asap. I recall some of the suggestions didn't work because of the design choices made by the upstream developer. I'll do what I can.

**Comment by Janik-Haag** [comment]
Thank you, take your time a few days more ore less won't matter at that point. If you cant resolve it the suggested way please comment why and mark it as resolved. 
That's a great contribution for your first one, feel free to make more :heart: 
You can also ping me in the future for reviews and merges and also if you have any questions please ask them.

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (michaelshmitty) [reply nixos/modules/services/web-apps/anuko-time-tracker.nix:292]: Upstream project only supports MySQL, unfortunately.

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (michaelshmitty) [reply nixos/modules/services/web-apps/anuko-time-tracker.nix:190]: I will stick to using the defaults that the upstream project provides. It's not critical for this PR to move forward, people are welcome to open an issue afterwards and I will address it as the package maintainer.

Comment (michaelshmitty) [reply nixos/modules/services/web-apps/anuko-time-tracker.nix:190]: > ~I will stick to using the defaults that the upstream project provides. It's not critical for this PR to move forward, people are welcome to open an issue afterwards and I will address it as the package maintainer.~

On second thought, looking at the next suggestion about using the default email address I agree that we're better off not setting a default. Definitely not for the sender address and then to stay consistent with that vision also not for the smtp port number.

File: nixos/modules/services/web-apps/anuko-time-tracker.nix
Comment (michaelshmitty) [reply nixos/modules/services/web-apps/anuko-time-tracker.nix]: Done, this was changed to use a passwordFile

**Comment by michaelshmitty** [comment]
@Janik-Haag thanks for pinging me on this. I've rebased, updated the package to the latest upstream version and addressed the suggestions by @SuperSandro2000 where I could. I've tested this on some of my servers and this seems to work well. I don't think there's much more that can be done but I'll be happy to maintain this and fix issues when they are opened. If this goes well, I'd be happy to look into maintaining more packages.

**Review by Janik-Haag — APPROVED** [review]

**Comment by Janik-Haag** [comment]
One small last nitpick, can you split the module in its own commit called `nixos/anuko-time-tracker: init` and the package init in `anuko-time-tracker: init at 1.22.10.5776`? After that once CI ran through successfully I'm happy to merge this. Thank you for your contribution, btw this looks great for your first module :heart: 
Happy to review your future PRs :D

**Comment by michaelshmitty** [comment]
> One small last nitpick, can you split the module in its own commit called `nixos/anuko-time-tracker: init` and the package init in `anuko-time-tracker: init at 1.22.10.5776`?

Sure, I'll do that. Thanks for your time!

**Review by Janik-Haag — APPROVED** [review]

File: pkgs/servers/web-apps/anuko-time-tracker/default.nix
Comment (Janik-Haag) [inline pkgs/servers/web-apps/anuko-time-tracker/default.nix:20]:
```suggestion
    hash = "sha256-ZRF9FFbntYY01JflCXkYZyXfyu/x7LNdyOB94UkVFR0=";
```
quite small nitpick, that I missed, doesn't change functionality but it's preferred when having a self describing hash like `sha256-....` :see_no_evil:

Comment (michaelshmitty) [reply pkgs/servers/web-apps/anuko-time-tracker/default.nix:20]: @Janik-Haag thanks for the tip! And don't worry, I don't see it as a nitpick. I prefer to be just as committed to detail. I'll fix it.

**Review by Janik-Haag — APPROVED** [review]

**Comment by Janik-Haag** [comment]
CI run was successful, and @michaelshmitty made the suggested changes and I couldn't find anything to improve anymore.

**Pull request merged**
